buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java

```diff
@@ -404,6 +409,19 @@ public void nextNodeToNextVersion() {
         node.start();
     }
 
+    public void upgradeNodeAndPluginToNextVersion(List<Provider<RegularFile>> plugins) {
```

Comment:
This is nice. It works for rolling upgrade/mixed cluster cases where we upgrade node by node. 
For full cluster restarts/restart upgrades we need another function to handle it for plugins.
Take a look at: 
```
public void goToNextVersion() 
```

Minor nit: I wish we didn't duplicate it but I dont see another way too :/

Reply:
I missed that! I will add that method as well since we are adding bwc support.

Reply:
Added support for full cluster restarts/restart upgrades. I have also tried to reduce duplication where possible.

Reply:
Awesome, we should be able to leverage restart cluster infra for snapshot upgrades. Essentially its taking snapshot one old version, store it in a repo, restart the cluster to new version and restore the snapshot. 
Lets open an issue to take care fo snapshot upgrades.

Reply:
Sure, opened an issue for tracking snapshot upgrade tests #1088.